### PR TITLE
Force memberaudit updates when a character applies

### LINKED ISSUE #17: Stale skill check data
It looks like the skills pulled for the skill check panel are out of date. I just verified that one of my alts does indeed meet all the skills in the skillset, but it's returning a red X. Character is Ara Khaspi (2118610448), for reference.

### PR DESCRIPTION
**Pull request by mome-borogove**
When a member applies, a memberaudit update is immediately queued for that character. Unless a staff member looks at the application within seconds of that action, this more or less solves the stale skill issue.

Because we're running MA 2.10, this is still a bit heavy-handed, since we run a *full* character update every time, instead of just pulling skills and corp/alliance. That can be easily fixed if MA is updated.

Fixes #17.

### CODE REVIEW

**Comment by lynkfox** [comment]
Players should not have the ability to do this, and it should not be automatic on application open. That will likely be a recipe for disaster 😞 

Instead, we should add to the skill check window pane the last time the skills were polled, and a button to do trigger it again. 


It doesn't need to be run on every application request

**Comment by mome-borogove** [comment]
I somewhat disagree, but I'm willing to be convinced.

> Players should not have the ability to do this, and it should not be automatic on application open. That will likely be a recipe for disaster 😞

Recall that applications are gated on a cooldown, both via rejection and withdrawal. It is *way* worse for WHC staff to be spammed with applications than it is for ESI to get a MA forced update with the same frequency. So if there's ever a situation where a player can cause issues due to MA update spam, they are causing far worse problems for staff, which needs to be addressed immediately. So I believe this is self-limiting in a way that never makes it an issue.

Also, the *total* number of MA force updates ever issued is necessarily equal to the total population of WHC plus those rejected. That's *not* a lot of people (when counted in MA updates per day), and it's spread over a long period of time.

> Instead, we should add to the skill check window pane the last time the skills were polled, and a button to do trigger it again.

The common case is that characters are likely only ever going to apply to WHC once, which means MA updates would have to come from some other source in Auth. Given that MA is not being updated currently, that's likely to be never.

> It doesn't need to be run on every application request

Given the state of MA, it really probably does. MA is currently useless. Most data is months or years out of date, if it exists at all, which means I suspect we'll be running a force update on every single app.


Basically, I'm arguing that a manual update will cause the same order of magnitude number of forced MA updates while *also* making staff click more buttons and spend more time. If I have a realistic choice between spending people time and computer time, I'm *always* going to choose the latter.


One additional thing: with Auth v4, we'll have a sharply reduced number of ESI requests for a WHC app compared to a normal MA update, since we don't need the whole character data pulled. The current old version of MA only has a single, monolithic force update task function, but the newer versions have finer granularity, which means we can just pick a couple relevant ones. (This applies evenly to both approaches, but it's worth knowing.)

**Comment by lynkfox** [comment]
The difference is that if a player re adds their character to auth
themselves, it also automatically pulls the latest. Now this isn't great
for mains, because removing that can cause some issues, but it's fine for
alts.

And I think we had a conversation about most players meet the skill
requirements on their main?

So I could see a compromise of pulling automatically on their main, but not
alts.

Also, you're still going to need a skill update button for staff  for any
app that takes longer than 24 hours to be reviewed - people are people,
they're going to 100% put in an app with just shy of the skill requirements
that will be finishing in  the next day or so.  I know there is a  warning
about the time gate but lol, people aren't going to think it applies to them

Lastly I'm more concerned about the opening days rather than in general.

That first day or two is going to see hundreds of apps opened, and if it
still is 3.8 then there will be 4x that amount of ESI requests. The
possibility of that causeing an automatic ban on the esi is not that far
fetched, especially since we rarely do that (tho perhaps Bates or Marn can
speak more to the likelihood here)

Maybe a decent compromise is automatic on mains, with a button for alts if
someone says they need it, and of course a staff button, and a celery task
put together to run once right before opening to fetch everyone in the pav
group? Celery update tasks like that have built in limiters I believe.

Another thing, that id be ok, is a time gate of its own on the member
check.  The self withdraw of application doesnt have a time gate I think?
Still don't want to punish people who accidentally push withdraw either.
So a time gate on the member audit of 24 hrs or something for that
situation makes a lot of sense to me

**Comment by mome-borogove** [comment]
> The difference is that if a player re adds their character to auth themselves, it also automatically pulls the latest. Now this isn't great for mains, because removing that can cause some issues, but it's fine for alts. And I think we had a conversation about most players meet the skill requirements on their main? So I could see a compromise of pulling automatically on their main, but not alts.

I'll check the difficulty of that. I suspect it's not terrible, and I'd be fine with it.

> Also, you're still going to need a skill update button for staff for any app that takes longer than 24 hours to be reviewed - people are people, they're going to 100% put in an app with just shy of the skill requirements that will be finishing in the next day or so. I know there is a warning about the time gate but lol, people aren't going to think it applies to them

I'll make sure to rosin up my tiny violin for those people.

Heck, I'm tempted to just add an auto-reject feature that *guarantees* those people get rejected instantly and then have to wait out the cooldown. Less work for the staff.

I'm not opposed to the update button on principle, I just happen to be opposed to most situations in which it's necessary---since those nearly always coincide with situations that cause a lot of headaches for CCs.

> Lastly I'm more concerned about the opening days rather than in general. That first day or two is going to see hundreds of apps opened, and if it still is 3.8 then there will be 4x that amount of ESI requests. The possibility of that causeing an automatic ban on the esi is not that far fetched, especially since we rarely do that (tho perhaps Bates or Marn can speak more to the likelihood here)

The initial push is coming regardless. So the only question we need to ask is whether automatic pulls are unnecessarily updating MA records. If an automatic update is just replicating what a manual one will do, then I don't want the manual one.

I do think I buy your point about alts not needing a pull. This brings us back to the whole "auth main / whc main" issue, since the only time that an alt *would* need a pull is if an auth-main doesn't meet the reqs and another of their characters does, which we have us in a situation where we're accepting an auth-main that *doesn't* meet the skills in order to allow a non-auth-main that we'd presumably be accepting to apply afterwards. That's a lot of added complexity that we don't really want (especially when one of those drops corp).

But all of that is a post-launch problem, because refactoring for an auth/whc main split is too much work right now. I'd rather get this off the ground.

> Maybe a decent compromise is automatic on mains, with a button for alts if someone says they need it, and of course a staff button...

Seems a bit overkill, but maybe fine. What's the "button for alts if someone says they need it" vs "a staff button"? You're not talking about a button on the user panel, right?

> ...and a celery task put together to run once right before opening to fetch everyone in the pav group?

If that's something you want, we can, but Pav is actually non-negligibly larger than the current active WHC population. There's a ton of people that dabbled and never left the group.

> Celery update tasks like that have built in limiters I believe.

If so, then the current code adheres to it. All we do is spawn a celery task with a reasonably high priority. That said, just from observing the launch behavior of celery, I don't think this is the case. Maybe they only rate limit after a certain number, though. I don't know.

> Another thing, that id be ok, is a time gate of its own on the member check. The self withdraw of application doesnt have a time gate I think?

Self-withdraw currently has a 2d (`WHCTOOLS_TRANSIENT_REJECT`) timer. So we're clear there. Again, I don't want any loopholes that can create arbitrary numbers of applications for staff to see, so I'm in total agreement with fixing them if we see any.

**Comment by mome-borogove** [comment]
> I could see a compromise of pulling automatically on their main, but not alts.

This was easy. It is done. Alts no longer trigger auto-updates, so the number of ESI calls is capped at the total number of humans applying to WHC, which is very manageable.

We can still go back and add manual triggers, last-update fields, buttons, and stuff, but the current status should be sufficient for launch IMO.

**Comment by lynkfox** [comment]
I'm good with that

**Comment by mome-borogove** [comment]
Going to go ahead and roll #46 in with this one, too, since it's mucking with the same apply-time MA pulls. Just needs a couple additional pieces stuck in there.

**Comment by mome-borogove** [comment]
Per Marn, using the EveCharacter corp/alliance info is kept up to date, so fixing #46 is straightforward.

**Review by Deadevilgrounds — APPROVED** [review]
I see that there are comments in there for MA 3.x
I will approve this PR and this will go onto the dev.eveuniversity.org instance which is using the latest MA. So feel free to make another PR if you want to use the latest MA features.
